### ISSUE TITLE
coredns: Pod Predicate MatchNodeSelector failed after 1.13.7 - 1.14.4

### ISSUE THREAD

**Issue by cann0nf0dder** [kubernetes/kubernetes#80982]
**What happened**:
After upgrade of the controllers/master nodes from 1.13.7 to 1.14.4 some coredns pods appear to have failed with "Pod Predicate MatchNodeSelector failed" on top of the running coredns pods

```
coredns-6789d857b9-6rlqx                   1/1     Running             0          179m    10.2.5.17      c8r005    <none>           <none>
coredns-6789d857b9-gsljz                   1/1     Running             0          3h11m   10.2.9.10      c8r002    <none>           <none>
coredns-6789d857b9-jb6vx                   1/1     Running             0          3h14m   10.2.15.11     c8r001    <none>           <none>
coredns-6789d857b9-jnc4l                   1/1     Running             1          18d     10.2.8.13      c8r003    <none>           <none>
coredns-6789d857b9-lqmhs                   1/1     Running             1          18d     10.2.2.10       c8r004    <none>           <none>
coredns-6789d857b9-s5lgs                   0/1     MatchNodeSelector   0          18d     <none>         c8r002    <none>           <none>
coredns-6789d857b9-vxkj5                   0/1     MatchNodeSelector   0          18d     <none>         c8r001    <none>           <none>

[root@b5n002 ~]# kubectl describe pod coredns-6789d857b9-s5lgs -n kube-system
Name:           coredns-6789d857b9-s5lgs
Namespace:      kube-system
Node:           c8r002/
Start Time:     Wed, 17 Jul 2019 15:17:29 +0000
Labels:         k8s-dns=coredns
                pod-template-hash=6789d857b9
Annotations:    cni.projectcalico.org/podIP: 10.2.9.8/32
Status:         Failed
Reason:         MatchNodeSelector
Message:        Pod Predicate MatchNodeSelector failed
IP:             

```
Logs from one of the controller for pod which ended up in this state (deleted manually)

```
user@c8r005 ~ $ sudo journalctl --since "5 hours ago" | grep coredns-6789d857b9-9h7t2
Aug 05 08:54:56 c8r005 kubelet-wrapper[47972]: W0805 08:54:56.807553   47972 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-9h7t2_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "c69e67fd43c0e59a5fe131a20a6a0334447cc9d4f879fb9a40f0b8f9b94071e1"
Aug 05 08:54:56 c8r005 kubelet-wrapper[47972]: W0805 08:54:56.824130   47972 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-9h7t2_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "16dae1a5f1d08bd80b37f7bdd9b4169f58c8a8b33ea8dc4280f471d77742c29f"
Aug 05 08:54:57 c8r005 kubelet-wrapper[47972]: 2019-08-05 08:54:57.265 [INFO][48103] k8s.go 470: Endpoint deletion will be handled by Kubernetes deletion of the Pod. ContainerID="16dae1a5f1d08bd80b37f7bdd9b4169f58c8a8b33ea8dc4280f471d77742c29f" endpoint=&v3.WorkloadEndpoint{TypeMeta:v1.TypeMeta{Kind:"WorkloadEndpoint", APIVersion:"projectcalico.org/v3"}, ObjectMeta:v1.ObjectMeta{Name:"c8r005-k8s-coredns--6789d857b9--9h7t2-eth0", GenerateName:"coredns-6789d857b9-", Namespace:"kube-system", SelfLink:"", UID:"c047bec6-a339-11e9-aac8-005056a1bf64", ResourceVersion:"125400761", Generation:0, CreationTimestamp:v1.Time{Time:time.Time{wall:0x0, ext:63698377204, loc:(*time.Location)(0x2341b20)}}, DeletionTimestamp:(*v1.Time)(nil), DeletionGracePeriodSeconds:(*int64)(nil), Labels:map[string]string{"k8s-dns":"coredns", "pod-template-hash":"6789d857b9", "projectcalico.org/namespace":"kube-system", "projectcalico.org/orchestrator":"k8s", "projectcalico.org/serviceaccount":"coredns"}, Annotations:map[string]string(nil), OwnerReferences:[]v1.OwnerReference(nil), Initializers:(*v1.Initializers)(nil), Finalizers:[]string(nil), ClusterName:""}, Spec:v3.WorkloadEndpointSpec{Orchestrator:"k8s", Workload:"", Node:"c8r005", ContainerID:"", Pod:"coredns-6789d857b9-9h7t2", Endpoint:"eth0", IPNetworks:[]string{"10.2.5.14/32"}, IPNATs:[]v3.IPNAT(nil), IPv4Gateway:"", IPv6Gateway:"", Profiles:[]string{"kns.kube-system", "ksa.kube-system.coredns"}, InterfaceName:"cali4e315e1ab16", MAC:"", Ports:[]v3.EndpointPort{v3.EndpointPort{Name:"dns", Protocol:numorstring.Protocol{Type:1, NumVal:0x0, StrVal:"UDP"}, Port:0x35}, v3.EndpointPort{Name:"dns-tcp", Protocol:numorstring.Protocol{Type:1, NumVal:0x0, StrVal:"TCP"}, Port:0x35}, v3.EndpointPort{Name:"metrics", Protocol:numorstring.Protocol{Type:1, NumVal:0x0, StrVal:"TCP"}, Port:0x23c1}}}}
Aug 05 08:54:57 c8r005 kubelet-wrapper[47972]: 2019-08-05 08:54:57.337 [INFO][48112] ipam.go 1164: Releasing all IPs with handle 'kube-system.coredns-6789d857b9-9h7t2'
Aug 05 08:54:57 c8r005 kubelet-wrapper[47972]: W0805 08:54:57.933755   47972 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-9h7t2_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "c69e67fd43c0e59a5fe131a20a6a0334447cc9d4f879fb9a40f0b8f9b94071e1"
Aug 05 11:42:38 c8r005 kubelet-wrapper[47972]: W0805 11:42:38.328067   47972 status_manager.go:501] Failed to update status for pod "coredns-6789d857b9-9h7t2_kube-system(c047bec6-a339-11e9-aac8-005056a1bf64)": failed to patch status "{}" for pod "kube-system"/"coredns-6789d857b9-9h7t2": pods "coredns-6789d857b9-9h7t2" not found
Aug 05 11:43:01 c8r005 kubelet-wrapper[47972]: 2019-08-05 11:43:01.179 [INFO][121760] ipam.go 1164: Releasing all IPs with handle 'kube-system.coredns-6789d857b9-9h7t2'
Aug 05 11:43:01 c8r005 kubelet-wrapper[47972]: W0805 11:43:01.184997   47972 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-9h7t2_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "c69e67fd43c0e59a5fe131a20a6a0334447cc9d4f879fb9a40f0b8f9b94071e1"
```

**What you expected to happen**:

I performed this upgrade on couple of clusters going from 1.13.7 to 1.14.4 and did not see this behaviour, all the system pods including coredns are running.

**Anything else we need to know?**:

We have recently removed some flags from kubelet.service since those are already set in API
Removed labels:
```
-  --node-labels=node-role.kubernetes.io/bastion \	
-  --register-with-taints=node-role.kubernetes.io/bastion=:NoSchedule
```

**Environment**:
```
[root@5n001 k8supgrade]# kubectl version
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.7", GitCommit:"4683545293d792934a7a7e12f2cc47d20b2dd01b", GitTreeState:"clean", BuildDate:"2019-06-06T01:46:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.4", GitCommit:"a87e9a978f65a8303aa9467537aa59c18122cbf9", GitTreeState:"clean", BuildDate:"2019-07-08T08:43:10Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}

NAME="Container Linux by CoreOS"
ID=coreos
VERSION=2135.5.0
VERSION_ID=2135.5.0
BUILD_ID=2019-07-01-1959
PRETTY_NAME="Container Linux by CoreOS 2135.5.0 (Rhyolite)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
```
- Network plugin and version (if this is a network-related bug): Calico 3.7.4
- Coredns 1.5.0

/sig node

**Comment by zouyee** [comment]
https://github.com/kubernetes/kubernetes/pull/80976

**Comment by cann0nf0dder** [comment]
FYI this has occurred when only one controller was being upgraded - NotReady, the remaining 4 controllers were available.

**Comment by cann0nf0dder** [comment]
I'm wondering if this could be because calico is potentially not ready before coredns pod? is there a restart priority that could be set to help? I'm using another cluster as example but this is the only log i see for the pod that gets in this state. 

`Aug 05 08:40:20 c8r001 kubelet-wrapper[91558]: W0805 08:40:20.329939   91558 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-vxkj5_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0"
`
`Aug 05 08:38:50 c8r001 env[893]: time="2019-08-05T08:38:50Z" level=info msg="shim reaped" id=04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0
Aug 05 08:38:50 c8r001 systemd[1]: var-lib-docker-containers-04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0-mounts-shm.mount: Succeeded.
Aug 05 08:38:50 c8r001 systemd[90560]: var-lib-docker-containers-04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0-mounts-shm.mount: Succeeded.
Aug 05 08:40:20 c8r001 kubelet-wrapper[91558]: W0805 08:40:20.329939   91558 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "coredns-6789d857b9-vxkj5_kube-system": CNI failed to retrieve network namespace path: cannot find network namespace for the terminated container "04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0"
Aug 05 08:40:25 c8r001 kubelet-wrapper[91558]: W0805 08:40:25.635568   91558 pod_container_deletor.go:75] Container "04fc9f6d753f6b8c9036022f87ef48c023740766e8bb1b6aebd5e1d629967ca0" not found in pod's containers"`

When i was going though all the logs, it feels to me that calico starts working much later around 08:40:46

**Comment by cann0nf0dder** [comment]
This could be a result of API server flapping during the upgrade when the pod is trying to deploy.

**Comment by cann0nf0dder** [comment]
Is someway this is related and it was fixed by #80173

**Comment by waelghaith** [comment]
Hello @cann0nf0dder 
Can you explain to us what the problem was and how can we solve it?!

**Comment by cann0nf0dder** [comment]
@waelghaith after we moved from 1.13 to 1.14 the issue resolved itself, it could have been related to  https://github.com/kubernetes/kubernetes/pull/80173 but i'm not 100%. It also was linked to the control plane flapping during upgrades https://github.com/kubernetes/kubernetes/issues/81286. We upgraded all of our clusters and did not experience any disruptions to the service since at least 1 out of 5 masters is available throughout the process. Not preferable but it worked out fine. If you are on 1.13 i'd suggest to upgrade to at least 1.14.6+
